contract.sol

```diff
@@ -0,0 +1,152 @@
+    }
+
+    // Function to claim rewards based on staked tokens
+    function claimReward(address rewardToken) external nonReentrant whenNotPaused {
+        Stake storage userStake = stakes[msg.sender][rewardToken];  // Access the user's stake for this reward token
+        if (userStake.amount == 0) revert NoStakeFound();  // Ensure the user has a stake
+        if (block.timestamp <= userStake.startTime + minStakingPeriod) revert CannotClaimRewardYet();  // Ensure the minimum staking period has passed
+
+        uint256 stakingDuration = block.timestamp - userStake.startTime;  // Calculate the staking duration
+        uint256 rewardAmount = (userStake.amount * rewardRate * stakingDuration) / 1e18;  // Calculate the reward amount
+
+        IERC20(rewardToken).transfer(msg.sender, rewardAmount);  // Transfer the reward tokens to the user
+
+        userStake.startTime = block.timestamp;  // Reset the stake start time
+
+        emit RewardClaimed(msg.sender, rewardAmount, rewardToken);  // Emit a RewardClaimed event
+    }
+
+    // Function for the contract owner to withdraw accumulated penalties
+    function withdrawPenalties() external onlyOwner nonReentrant whenNotPaused {
+        uint256 amountToWithdraw = accumulatedPenalties;  // Get the total accumulated penalties
+        if (amountToWithdraw == 0) revert NoPenaltiesToWithdraw();  // Ensure there are penalties to withdraw
+
+        accumulatedPenalties = 0;  // Reset the accumulated penalties
+        stakingToken.transfer(owner(), amountToWithdraw);  // Transfer the penalties to the owner
+
+        emit PenaltiesWithdrawn(owner(), amountToWithdraw);  // Emit a PenaltiesWithdrawn event
+    }
+
+    // Function to add a new reward token
+    function addRewardToken(address rewardToken) external onlyOwner {
+        if (isRewardToken[rewardToken]) revert InvalidRewardToken();  // Ensure the token is not already a reward token
+        rewardTokens.push(rewardToken);  // Add the token to the reward tokens list
+        isRewardToken[rewardToken] = true;  // Mark the token as a valid reward token
+    }
+
+    // Function to set the minimum staking period
+    function setMinStakingPeriod(uint256 period) external onlyOwner {
+        minStakingPeriod = period;
+    }
+
+    // Function to set the early withdrawal penalty percentage
+    function setEarlyWithdrawalPenalty(uint256 penalty) external onlyOwner {
+        earlyWithdrawalPenalty = penalty * 1e18;
+    }
+
+    // Function to set the reward rate
+    function setRewardRate(uint256 rate) external onlyOwner {
+        rewardRate = rate * 1e18;
+    }
+
+    // Function to get the list of reward tokens
+    function getRewardTokens() external view returns (address[] memory) {
+        return rewardTokens;
+    }
+
+    // Function to get the staked amount of a user for a specific reward token
+    function getStakedAmount(address user, address rewardToken) external view returns (uint256) {
```

Comment:
function for getting all positions where user staked? 
user won't search through all 1000 reward tokens and input in this function to get his rewards.
Please think from user's perspective.